README.rst
====
xbin
====


.. image:: https://img.shields.io/pypi/v/xbin.svg
        :target: https://pypi.python.org/pypi/xbin

.. image:: https://img.shields.io/travis/willsheffler/xbin.svg
        :target: https://travis-ci.org/willsheffler/xbin

.. image:: https://readthedocs.org/projects/xbin/badge/?version=latest
        :target: https://xbin.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Fancy Spatial Binning (Hashing)


* Free software: Apache Software License 2.0
* Documentation: https://xbin.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.5 (2018-07-10)
------------------

* add numba support

0.1.5 (2018-03-28)
------------------

* maybe working, covrad seems reasonable-ish

0.1.2 (2018-03-07)
------------------

* add basic bt24cell stuff
* First release on PyPI.


tests/test_xbin.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `xbin` package."""

import pytest


from xbin import xbin


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


xbin/tests/test_xbin.py
from xbin import *
import homog
import pytest
from traitlets import TraitError
from hypothesis import given, settings
import hypothesis.strategies as hs
import hypothesis.extra.numpy as hn

try:
    import numba
    only_if_numba = lambda f: f
except ImportError:
    import pytest
    only_if_numba = pytest.mark.skip

# print(('!' * 80 + '\n') * 10, end='')
MAX_EXAMPLES = 100


def test_get_half_bt24cell_face_basic():
    assert half_bt24cell_faces.shape == (24, 4)
    assert np.all(
        homog.quat.quat_to_upper_half(half_bt24cell_faces) ==
        half_bt24cell_faces)
    assert np.min(np.linalg.norm(half_bt24cell_faces, axis=-1)) == 1
    assert np.max(np.linalg.norm(half_bt24cell_faces, axis=-1)) == 1
    assert half_bt24cell_face([0, 0, 0, 1]) == 3
    assert half_bt24cell_face([1, -1, 1, 1]) == 8
    assert half_bt24cell_face([0.5, 0.5, 0, 0]) == 12
    assert half_bt24cell_face([-0.5, 0, 0, 0.5]) == 17
    assert np.all(half_bt24cell_face(half_bt24cell_faces) == np.arange(24))
    assert np.all(half_bt24cell_face(-half_bt24cell_faces) == np.arange(24))
    q = homog.quat.rand_quat((7, 6, 5))
    x = half_bt24cell_face(q)
    assert x.shape == (7, 6, 5)
    assert np.all(x >= 0)
    assert np.all(x <= 23)


@only_if_numba
def test_numba_half_bt24cell_face():
    assert numba_half_bt24cell_face(np.array([0, 0, 0, 1], dtype='f8')) == 3
    assert numba_half_bt24cell_face(np.array([1, -1, 1, 1], dtype='f8')) == 8
    assert numba_half_bt24cell_face(np.array([0.5, 0.5, 0, 0],
                                             dtype='f8')) == 12
    assert numba_half_bt24cell_face(np.array([-0.5, 0, 0, 0.5],
                                             dtype='f8')) == 17


def test_half_bt24cell_geom():
    cellquatang = 2 * np.arccos(
        np.clip(
            np.sum(half_bt24cell_faces * half_bt24cell_faces[:, None], -1), -1,
            1))
    cellquatang = np.minimum(cellquatang, 2 * np.pi - cellquatang)
    cellcen = homog.quat.quat_to_xform(half_bt24cell_faces)
    cellcenang = homog.angle_of(homog.hinv(cellcen) @ cellcen[:, None])

    cellcenanground = np.round(cellcenang * 180 / np.pi, 5)
    # print(np.unique(cellcenanground))

    assert np.allclose(cellcenang, cellquatang)
    # np.fill_diagonal(cellcenang, 10)

    q = homog.quat.rand_quat((1000, ))
    qang = 2 * np.arccos(
        np.clip(np.abs(np.sum(q * half_bt24cell_faces[:, None], -1)), -1, 1))
    minqang = np.min(qang, 0)
    # print('Qang to closest',
    #        np.percentile(minqang * 180 / np.pi, np.arange(0, 101, 10)))
    assert np.max(minqang) < 62.81 / 180 * np.pi

    x = homog.rand_xform(1000)
    cellxform = homog.quat.quat_to_xform(half_bt24cell_faces)
    xang = homog.angle_of(x @ homog.hinv(cellxform)[:, None])
    minxang = np.min(xang, 0)
    # print('Xang to closest:',
    # np.percentile(minxang * 180 / np.pi, np.arange(0, 101, 10)))
    assert np.max(minqang) < 62.81 / 180 * np.pi


def test_XformBinner_init():
    with pytest.raises(ValueError):
        XformBinner(cart_resl=-1)
    with pytest.raises(ValueError):
        XformBinner(ori_resl=-1)
    with pytest.raises(ValueError):
        XformBinner(cart_bound=-1)
    assert XformBinner().cart_resl == 1
    assert XformBinner(cart_resl=2).cart_resl == 2


def test_xform_to_f6():
    x = homog.rand_xform(24000)
    face, f6 = xform_to_f6(x)
    assert np.all(face >= 0)
    assert np.all(face < 24)
    assert np.all(f6[..., 3:] >= 0)
    assert np.all(f6[..., 3:] <= 1)
    counts = np.sum(face == np.arange(24)[..., None], axis=-1)
    assert np.all(counts < 1200)
    assert np.all(counts > 800)

    xbt24 = homog.quat.quat_to_xform(half_bt24cell_faces)
    assert np.all(homog.is_homog_xform(xbt24))
    qbt24 = homog.quat.xform_to_quat(xbt24)
    assert np.all(homog.quat.is_valid_quat_rot(qbt24))
    assert np.allclose(qbt24, half_bt24cell_faces)
    face = half_bt24cell_face(qbt24)
    assert np.all(face == np.arange(24))


@only_if_numba
def test_numba_to_face_0():
    for i in range(1000):
        q = homog.quat.rand_quat()


@only_if_numba
def test_numba_xform_to_f6():
    for i in range(1000):
        x = homog.rand_xform(1)
        face_0, f6_0 = xform_to_f6(x)
        face_1, f6_1 = numba_xform_to_f6(x[0])
        assert face_0 == face_1
        if not np.allclose(f6_0, f6_1):
            print(i, face_0, face_1)
            print(f6_0)
            print(f6_1)
            print(hg.quat.rot_to_quat(x))
            print(hg.quat.numba_rot_to_quat(x[0]))
            assert 0


@hs.composite
def Quaternions(draw, shape=()):
    N = int(np.prod(shape))
    quat = draw(hn.arrays(np.float, (0, 4), elements=hs.floats(-1, 1)))
    while quat.shape[0] < N:
        qbox = draw(
            hn.arrays(
                dtype=np.float,
                shape=(2 * (N - quat.shape[0]), 4),
                elements=hs.floats(-1, 1)))
        norm = np.linalg.norm(qbox, axis=-1)
        qbox[norm == 0] = [1, 0, 0, 0]
        norm[norm == 0] = 1
        ok = (0.01 < norm) * (norm <= 1.0)
        qboxok = qbox[ok] / norm[ok, np.newaxis]
        quat = np.concatenate([quat, qboxok])
    return quat[:N].reshape(shape + (4, ))


@hs.composite
def Xforms(draw, shape=()):
    quat = draw(Quaternions(shape))
    x = homog.quat.quat_to_xform(quat)
    trns = draw(
        hn.arrays(np.float, shape + (3, ), elements=hs.floats(-10, 10)))
    x[..., :3, 3] = trns
    return x


@pytest.mark.skip
@given(hn.arrays(np.float, (1000, 6), elements=hs.floats(0, 1)))
@settings(max_examples=MAX_EXAMPLES)
def test_hypothesis_f6_to_xform_invertibility_face0(f6):
    assert np.all(f6[..., 3:] >= 0)
    assert np.all(f6[..., 3:] <= 1)
    f6 = f6[is_contained_in_bt24cell_face0(f6)]
    x = f6_to_xform(0, f6)
    assert np.allclose(x[..., :3, 3], f6[..., :3])
    face, f6b = xform_to_f6(x)
    assert np.sum(face != 0) < 10
    assert np.allclose(f6[face == 0], f6b[face == 0])


@given(Xforms(shape=(100, )))
@settings(max_examples=MAX_EXAMPLES)
def test_hypothesis_f6_to_xform_invertibility(xforms):
    face, f6 = xform_to_f6(xforms)
    assert np.all((0 <= face) * (face < 24))
    assert np.all(f6[..., 3:] >= 0)
    assert np.all(f6[..., 3:] <= 1)
    xhat = f6_to_xform(face, f6)
    assert np.allclose(xhat, xforms)


def test_f6_to_quat():

    assert np.allclose(
        xbin._f6_to_quat(np.array([0, 0, 0, 0.0, 0.0, 0.0])),
        [0.81251992, -0.33655677, -0.33655677, -0.33655677])
    assert np.allclose(
        xbin._f6_to_quat(np.array([0, 0, 0, 0.5, 0.5, 0.5])), [1, 0, 0, 0])
    assert np.allclose(
        xbin._f6_to_quat(np.array([0, 0, 0, 1.0, 1.0, 1.0])),
        [0.81251992, 0.33655677, 0.33655677, 0.33655677])


def test_f6_to_xform_invertibility_face0():
    nexamples = 100000
    f6 = np.random.rand(nexamples, 6)
    f6 = f6[is_contained_in_bt24cell_face0(f6)]
    assert f6.shape[0] > nexamples / 10
    assert np.all(f6[..., 3:] >= 0)
    assert np.all(f6[..., 3:] <= 1)
    x = f6_to_xform(0, f6)
    assert np.allclose(x[..., :3, 3], f6[..., :3])
    face, f6b = xform_to_f6(x)
    assert np.all(np.min(f6b[..., 3:], axis=0) < 0.1)
    assert np.all(np.max(f6b[..., 3:], axis=0) > 0.9)
    assert np.all(face == 0)
    assert np.allclose(f6, f6b)


def test_f6_to_xform_invertibility():
    nexamples = 100000
    xforms = homog.rand_xform((nexamples, ))
    face, f6 = xform_to_f6(xforms)
    assert np.all(np.min(f6[..., 3:], axis=0) < 0.1)
    assert np.all(np.max(f6[..., 3:], axis=0) > 0.9)
    assert np.all((0 <= face) * (face < 24))
    assert np.all(f6[..., 3:] >= 0)
    assert np.all(f6[..., 3:] <= 1)
    xhat = f6_to_xform(face, f6)
    assert np.allclose(xhat, xforms)


def test_too_many_bins():
    XformBinner(0.1, 3)
    with pytest.raises(ValueError):
        XformBinner(0.01, 3)
    with pytest.raises(ValueError):
        XformBinner(0.01, 10)


def test_XformBinner_covrad():
    niter = 30
    nsamp = 10000
    for i in range(niter):
        cart_resl = np.random.rand() * 10 + 0.1
        ori_resl = np.random.rand() * 50 + 2.5
        xforms = homog.rand_xform(nsamp)
        xb = XformBinner(cart_resl, ori_resl)
        idx = xb.get_bin_index(xforms)
        cen, f6 = xb.get_bin_center(idx, debug=True)
        cart_dist = np.linalg.norm(
            xforms[..., :3, 3] - cen[..., :3, 3], axis=-1)
        ori_dist = homog.angle_of(homog.hinv(cen) @ xforms)
        # if not np.all(cart_dist < cart_resl):
        # print('ori_resl', ori_resl, 'nside:', xb.ori_nside,
        # 'max(cart_dist):', np.max(cart_dist), cart_resl)
        # if not np.all(cart_dist < cart_resl):
        # print('ori_resl', ori_resl, 'nside:', xb.ori_nside,
        # 'max(ori_dist):', np.max(ori_dist))
        assert np.sum(cart_dist < cart_resl) > nsamp * 0.99
        assert np.sum(ori_dist < ori_resl / 180 * np.pi) > nsamp * 0.99


def test_xbinner_bcc6_alignment():
    cart_resl = 512
    ori_resl = 63
    xb = XformBinner(cart_resl, ori_resl)
    # print(xb.bcc6.sizes)
    bcc6_cen = xb.bcc6.get_bin_center(np.arange(len(xb.bcc6)))
    # print(bcc6_cen.shape)
    inbounds = ((xb.bcc6.lower < bcc6_cen) * (bcc6_cen < xb.bcc6.upper))
    inbounds = np.prod(inbounds, axis=-1) == 1
    # print(np.sum(inbounds))
    # print(bcc6_cen[inbounds])
    # print('nside', xb.bcc6.nside)
    assert np.all(np.abs(inbounds[0]) == 0)


@only_if_numba
def test_numba_xbin():
    cart_resl = 1.0
    ori_resl = 10.0
    xbinner = XformBinner(cart_resl, ori_resl)
    xbin_indexer = numba_xbin_indexer(cart_resl, ori_resl)

    for i in range(100):
        x = homog.rand_xform(1)
        assert xbinner.get_bin_index(x) == xbin_indexer(x[0])

    xbin_indexer = gu_xbin_indexer(cart_resl, ori_resl)

    x = homog.rand_xform(100000, cart_sd=100)
    idx0 = xbinner.get_bin_index(x)
    idx1 = xbin_indexer(x)
    assert np.all(idx0 == idx1)

scripts/estimate_covrad.py
from contextlib import redirect_stdout
import numpy as np
import sys, os
import homog
with redirect_stdout(open(os.devnull,'w')):
    from xbin import *

ori_nside = int(sys.argv[1])
Nsamp = 1000000
Nptiles = 101
ptiles = np.concatenate([np.arange(Nptiles),
                         100 - 10**np.linspace(2, -6, Nptiles)])
estimate = np.zeros_like(ptiles)
coherent = np.zeros_like(ptiles)

while True:
    xforms = homog.rand_xform(Nsamp)
    xb = XformBinner(ori_nside=ori_nside)
    idx = xb.get_bin_index(xforms)
    cen = xb.get_bin_center(idx)
    ori_dist = homog.angle_of(homog.hinv(cen) @ xforms)
    ehat = np.percentile(ori_dist, ptiles)
    old = estimate
    estimate = np.maximum(estimate, ehat)
    if np.any(old != estimate):
        print('independent', Nsamp, ori_nside,
              str(estimate.tolist()).replace(',','')[1:-1])
        sys.stdout.flush()
    if np.sum(ehat) > np.sum(coherent):
        coherent = ehat
        print('coherent', Nsamp, ori_nside,
              str(coherent.tolist()).replace(',','')[1:-1])
        sys.stdout.flush()


xbin/xbin.py
import sys
# sys.path.insert(0, '/home/sheffler/src/homog/')
# print(('!' * 80 + '\n') * 10, end='')

import functools as ft
import numpy as np
from bcc import BCC
try:
    from bcc import numba_bcc_indexer
    HAVE_NUMBA = True
except ImportError:
    HAVE_NUMBA = False
import homog as hg
from homog.util import jit, guvec, float64, int64
import operator

if HAVE_NUMBA:

    def gu_xbin_indexer(*args, **kw):
        func = _kernel_xbin_indexer(*args, **kw)
        return guvec([(float64[:, :], int64[:])], '(n,n)->()', func)

    def numba_xbin_indexer(*args, **kw):
        func0 = _kernel_xbin_indexer(*args, **kw)

        @jit
        def func(value):
            res = np.empty(1, dtype=np.int64)
            func0(value, res)
            return res[0]

        return func

    def _kernel_xbin_indexer(cart_resl,
                             ori_resl,
                             cart_bound=512.0,
                             ori_nside=None):
        spec = xbin_make_spec(cart_resl, ori_resl, cart_bound, ori_nside)
        bcc6_indexer = numba_bcc_indexer(**spec)
        assert bcc6_indexer

        @jit
        def func(xform, index):
            face, f6 = numba_xform_to_f6(xform)
            idx = bcc6_indexer(f6)
            index[0] = np.bitwise_or(np.left_shift(face, 58), idx)

        return func


class XformBinner:
    def __init__(self,
                 cart_resl=1,
                 ori_resl=15,
                 cart_bound=512.0,
                 ori_nside=None):
        self.cart_resl = cart_resl
        self.ori_resl = ori_resl
        self.cart_bound = cart_bound
        spec = xbin_make_spec(cart_resl, ori_resl, cart_bound, ori_nside)
        self.bcc6 = BCC(**spec)

    def get_bin_index(self, xform, debug=False):
        face, f6 = xform_to_f6(xform)
        # print('get_bin_index face', face)
        idx = self.bcc6.get_bin_index(f6)
        index = np.bitwise_or(np.left_shift(face, 58), idx)
        if debug: index = index, f6
        return index

    def get_bin_center(self, index, debug=False):
        face = np.right_shift(index, 58)
        if not np.all((0 <= face) * (face < 24)):
            print('get_bin_center face', face[(0 > face) + (face > 24)])
            assert np.all((0 <= face) * (face < 24))
        bcc_key = np.right_shift(np.left_shift(index, 6), 6)
        assert np.all(np.right_shift(bcc_key, 58) == 0)
        f6 = self.bcc6.get_bin_center(bcc_key)
        center = f6_to_xform(face, f6)
        if debug: center = center, f6
        return center


def xbin_make_spec(cart_resl=1, ori_resl=15, cart_bound=512.0, ori_nside=None):
    if cart_resl <= 0:
        raise ValueError('cart_resl must be > 0, not ' + str(cart_resl))
    if ori_resl <= 0:
        raise ValueError('ori_resl must be > 0, not ' + str(ori_resl))
    if cart_bound <= 0:
        raise ValueError('cart_bound must be > 0, not ' + str(cart_bound))
    if not ori_nside:
        ori_nside = int(np.sum(_xform_binner_covrad >= ori_resl) + 1)
    # 1.7 is empirically determined hack...
    sizes = ([int(1.5 * cart_bound / cart_resl)] * 3 + [int(ori_nside)] * 3)
    totsize = ft.reduce(operator.mul, [_ + 2 for _ in sizes], 1)
    if totsize > 2**58:
        raise ValueError('too many bins, increase cart_resl and/or ori_resl,'
                         ' or reduce cart_bound')
    return dict(
        sizes=sizes,
        lower=[-cart_bound] * 3 + [0] * 3,
        upper=[cart_bound] * 3 + [1] * 3)


bt24cell_width = 2 * (np.sqrt(2) - 1)
bt24cell_width_diagonal = 0.696923425058676
_R22 = np.sqrt(2) / 2
half_bt24cell_faces = np.array([
    [1.000, 0.000, 0.000, 0.000],  # 0
    [0.000, 1.000, 0.000, 0.000],  # 1
    [0.000, 0.000, 1.000, 0.000],  # 2
    [0.000, 0.000, 0.000, 1.000],  # 3
    [+0.50, +0.50, +0.50, +0.50],  # 4
    [+0.50, +0.50, +0.50, -0.50],  # 5
    [+0.50, +0.50, -0.50, +0.50],  # 6
    [+0.50, +0.50, -0.50, -0.50],  # 7
    [+0.50, -0.50, +0.50, +0.50],  # 8
    [+0.50, -0.50, +0.50, -0.50],  # 9
    [+0.50, -0.50, -0.50, +0.50],  # 10
    [+0.50, -0.50, -0.50, -0.50],  # 11
    [+_R22, +_R22, 0.000, 0.000],  # 12
    [+_R22, -_R22, 0.000, 0.000],  # 13
    [+_R22, 0.000, +_R22, 0.000],  # 14
    [+_R22, 0.000, -_R22, 0.000],  # 15
    [+_R22, 0.000, 0.000, +_R22],  # 16
    [+_R22, 0.000, 0.000, -_R22],  # 17
    [0.000, +_R22, +_R22, 0.000],  # 18
    [0.000, +_R22, -_R22, 0.000],  # 19
    [0.000, +_R22, 0.000, +_R22],  # 20
    [0.000, +_R22, 0.000, -_R22],  # 21
    [0.000, 0.000, +_R22, +_R22],  # 22
    [0.000, 0.000, +_R22, -_R22],  # 23
])
half_bt24cell_faces.flags.writeable = False


def half_bt24cell_face(quat):
    # maybe faster to do the c++ stuff:
    # template <class V4, class Index>
    # void get_cell_bt24cell_half(V4 const &quat, Index &cell) {
    #   typedef typename V4::Scalar Float;
    #   V4 const quat_pos = quat.cwiseAbs();
    #   V4 tmpv = quat_pos;

    #   Float hyperface_dist;  // dist to closest face
    #   Index hyperface_axis;  // closest hyperface-pair
    #   Float edge_dist;       // dist to closest edge
    #   Index edge_axis_1;     // first axis of closest edge
    #   Index edge_axis_2;     // second axis of closest edge
    #   Float corner_dist;     // dist to closest corner

    #   // std::cout << quat_pos.transpose() << std::endl;
    #   numeric::max2(quat_pos, hyperface_dist, edge_dist, hyperface_axis,
    #                 edge_axis_2);
    #   edge_dist = sqrt(2) / 2 * (hyperface_dist + edge_dist);
    #   corner_dist = quat_pos.sum() / 2;
    #   // std::cout << hyperface_axis << " " << edge_axis_2 << std::endl;
    #   edge_axis_1 = hyperface_axis < edge_axis_2 ? hyperface_axis : edge_axis_2;
    #   edge_axis_2 = hyperface_axis < edge_axis_2 ? edge_axis_2 : hyperface_axis;
    #   assert(edge_axis_1 < edge_axis_2);

    #   // cell if closest if of form 1000 (add 4 if negative)
    # Index facecell = hyperface_axis;  // | (quat[hyperface_axis]<0 ? 4 :
    # 0);

    #   // cell if closest is of form 1111, bitwise by ( < 0)
    #   Index bit0 = quat[0] < 0;
    #   Index bit1 = quat[1] < 0;
    #   Index bit2 = quat[2] < 0;
    #   Index cornercell = quat[3] > 0 ? bit0 | bit1 << 1 | bit2 << 2
    #                                  : (!bit0) | (!bit1) << 1 | (!bit2) << 2;

    #   // cell if closest is of form 1100
    #   Index perm_shift[3][4] = {{9, 0, 1, 2}, {0, 9, 3, 4}, {1, 3, 9, 5}};
    #   Index sign_shift = (quat[edge_axis_1] < 0 != quat[edge_axis_2] < 0) * 1 * 6;
    #   Index edgecell = sign_shift + perm_shift[edge_axis_1][edge_axis_2];

    #   // pick case 1000 1111 1100 without if statements
    #   Index swtch;
    #   util::SimpleArray<3, Float>(hyperface_dist, corner_dist, edge_dist)
    #       .maxCoeff(&swtch);
    #   cell = swtch == 0 ? facecell : (swtch == 1 ? cornercell + 4 : edgecell + 12);
    #   // this is slower !?!
    #   // Float mx = std::max(std::max(hyperface_dist,corner_dist),edge_dist);
    #   // cell2[i] = hyperface_dist==mx ? facecell : (corner_dist==mx ? cornercell+8
    #   // : edgecell+24);
    # }
    quat = np.asarray(quat)
    fullaxes = (slice(None), ) + (np.newaxis, ) * (quat.ndim - 1)
    hf = half_bt24cell_faces[fullaxes]
    dots = abs(np.sum(quat[np.newaxis] * hf, axis=-1))
    return np.argmax(dots, axis=0)


@hg.util.jit
def numba_half_bt24cell_face(quat):
    dots = np.abs(half_bt24cell_faces @ quat)
    return np.argmax(dots)


def to_face_0(quat):
    face = half_bt24cell_face(quat)
    to_0 = half_bt24cell_faces[face].copy()
    to_0[..., 0] *= -1.0  # inverse
    q0 = hg.quat.quat_multiply(to_0, quat)
    return face, hg.quat.quat_to_upper_half(q0)


@hg.util.jit
def numba_to_face_0(quat):
    face = numba_half_bt24cell_face(quat)
    to_0 = half_bt24cell_faces[face].copy()
    to_0[0] *= -1.0  # inverse
    q0 = hg.quat.numba_quat_multiply(to_0, quat)
    q0 = hg.quat.numba_quat_to_upper_half(q0)
    return face, q0


_xform_binner_covrad = np.array([
    68.907736, 38.232608, 25.860598, 19.391913, 15.461347, 13.152186,
    11.309651, 9.849257, 8.653823, 7.778357, 7.044655, 6.504872, 5.971447,
    5.631019, 5.240999, 4.830578, 4.609091, 4.373091, 4.113840, 3.855201,
    3.726551, 3.529638, 3.377918, 3.262408, 3.171638, 2.978760, 2.879508,
    2.851197, 2.672736, 2.628641, 2.515808, 2.448288
])
_xform_binner_covrad.flags.writeable = False


def _f6_to_quat(f6):
    quat = np.empty(f6.shape[:-1] + (4, ))
    quat[..., 0] = 1.0
    quat[..., 1:] = bt24cell_width * (np.clip(f6[..., 3:6], 0, 1) - 0.5)
    return quat / np.linalg.norm(quat, axis=-1)[..., np.newaxis]


def is_contained_in_bt24cell_face0(f6):
    return half_bt24cell_face(_f6_to_quat(f6)) == 0


def f6_to_xform(face, f6):
    quat = _f6_to_quat(f6)
    quat = hg.quat.quat_multiply(half_bt24cell_faces[face], quat)
    xform = hg.quat.quat_to_xform(quat)
    xform[..., :3, 3] = f6[..., :3]
    return xform


def xform_to_f6(xform):
    quat = hg.quat.rot_to_quat(xform)
    face, q0 = to_face_0(quat)
    f6 = np.empty(q0.shape[:-1] + (6, ))
    f6[..., :3] = xform[..., :3, 3]
    # TODO (w,x,y,z) why divide by xyz by w? in cell0, w close to 1.0?
    ori_params = 0.5 + q0[..., 1:4] / q0[..., 0, None] / bt24cell_width
    f6[..., 3:6] = np.clip(ori_params, 0, 1)
    return face, f6


@hg.util.jit
def numba_xform_to_f6(xform):
    quat = hg.quat.numba_rot_to_quat(xform)
    face, q0 = numba_to_face_0(quat)
    f6 = np.empty((6, ))
    f6[:3] = xform[:3, 3]
    # TODO (w,x,y,z) why divide by xyz by w? in cell0, w close to 1.0?
    ori_params = 0.5 + q0[1:4] / q0[0] / bt24cell_width
    f6[3:6] = np.minimum(1, np.maximum(0, ori_params))
    return face, f6
